views.go

```diff
@@ -698,75 +538,54 @@ func (lk *logKeeper) viewBucketLogs(r *http.Request, buildID string, testID stri
 		logLinesErr error
 	)
 
-	// We request extra data we need in goroutines in parallel to save latency,
-	// and fetch the build itself on the main thread. If there is an issue with
-	// fetching the build then we don't wait for the goroutines to finish before
-	// falling back to the DB.
-	backgroundContext, backgroundCancel := context.WithCancel(r.Context())
-	wg.Add(2)
+	wg.Add(3)
+	go func() {
+		defer recovery.LogStackTraceAndContinue("finding build from bucket")
+		defer wg.Done()
+
+		build, buildErr = lk.opts.Bucket.FindBuildByID(r.Context(), buildID)
+	}()
 	go func() {
 		defer recovery.LogStackTraceAndContinue("finding test for build from bucket")
 		defer wg.Done()
 
 		if testID == "" {
 			return
 		}
-		test, testErr = lk.opts.Bucket.FindTestByID(backgroundContext, buildID, testID)
+		test, testErr = lk.opts.Bucket.FindTestByID(r.Context(), buildID, testID)
 	}()
 	go func() {
 		defer recovery.LogStackTraceAndContinue("downloading log lines from bucket")
 		defer wg.Done()
 
-		logLines, logLinesErr = lk.opts.Bucket.DownloadLogLines(backgroundContext, buildID, testID)
+		logLines, logLinesErr = lk.opts.Bucket.DownloadLogLines(r.Context(), buildID, testID)
 	}()
+	wg.Wait()
 
-	build, buildErr = lk.opts.Bucket.FindBuildByID(r.Context(), buildID)
-	if pail.IsKeyNotFoundError(buildErr) {
-		backgroundCancel()
-		if shouldFallBack {
-			grip.Info(message.Fields{
-				"message":   "fallback to db",
-				"operation": "viewBucketLogs",
-				"build_id":  buildID,
-				"test_id":   testID,
-			})
-			if testID == "" {
-				return lk.viewAllDBLogs(r, buildID)
-			} else {
-				return lk.viewDBTestLogs(r, buildID, testID)
-			}
-		} else {
-			return nil, &apiError{Err: "finding build", code: http.StatusNotFound}
-		}
-	}
 	if buildErr != nil {
-		backgroundCancel()
 		lk.logErrorf(r, "finding build '%s': %v", buildID, buildErr)
-		return nil, &apiError{Err: "finding build", code: http.StatusInternalServerError}
+		return nil, &apiError{Err: "finding build", code: http.StatusInternalServerError}, false
 	}
-
-	wg.Wait()
-	if testID != "" && pail.IsKeyNotFoundError(testErr) {
-		backgroundCancel()
-		return nil, &apiError{Err: "test not found", code: http.StatusNotFound}
+	if build == nil {
+		return nil, &apiError{Err: "build not found", code: http.StatusNotFound}, true
```

Comment:
was it not ideal to cancel the context when the build isn't found?

Reply:
I don't think we get much benefit from it since we are running these in parallel and if the build isn't found the other goroutines should return just as quick since there shouldn't be any test metadata or log lines for a build that does not exist.